src/JakobSteinn/HTMLFormValidator/Validator.php

```diff
@@ -0,0 +1,174 @@
+<?php
+namespace JakobSteinn\HTMLFormValidator;
+
+use \DOMElement;
+use \DOMDocument;
+use JakobSteinn\HTMLFormValidator\RuleProviders\URLRuleProvider;
+use JakobSteinn\HTMLFormValidator\RuleProviders\EmailRuleProvider;
+use JakobSteinn\HTMLFormValidator\Exceptions\UnknownValidationRule;
+
+class Validator
+{
+    /**
+     * The document that will be searched for a form
+     * 
+     * @var DOMDocument
+     */
+    protected $formDocument;
+
+    /**
+     * The rule providers used to validate input
+     * 
+     * @var JakobSteinn\HTMLFormValidator\RuleProviders\AbstractRuleProvider
+     */
+    protected $ruleProviders = [
+        'email'        => EmailRuleProvider::class,
+        'url'            => URLRuleProvider::class,
+    ];
+
+    /**
+     * List of rules to check
+     * 
+     * @var array
+     */
+    protected $rules = [];
+
+    /**
+     * List of errors from validation
+     *
+     * @var array
+     */
+    protected $errors = [];
+
+    /**
+     * Constructor
+     * 
+     * @param string $formHTML
+     */
+    public function __construct($formHTML)
+    {
+        $this->formDocument = new DOMDocument();
+        $this->formDocument->loadHTML($formHTML);
+        $this->parseForRules();
+    }
+
+    /**
+     * Loop the form data through form rules
+     * 
+     * @param  array  $formData		The associated form data
+     * @throws JakobSteinn\HTMLFormValidator\Exceptions\UnknownValidationRule
+     * @return JakobSteinn\HTMLFormValidator\Validator
+     */
+    public function validate(array $formData)
+    {
+        foreach ($this->rules as $field => $fieldRule) {
+            $currentFieldData = $formData[$field];
+
+            if (! class_exists($this->ruleProviders[$fieldRule])) {
+                throw new UnknownValidationRule('No rule defined for: ' + $field);
+            }
+
+            $currentRule = (new \ReflectionClass($this->ruleProviders[$fieldRule]))
+                ->newInstance()->run($field, $currentFieldData);
+
+            if ($currentRule->hasPassed) {
+                continue;
+            }
+
+            $this->errors[$currentRule->field] = $currentRule->message;
+        }
+
+        return $this;
+    }
+
+    /**
+     * Determine whether the validation has passed, returns true
+     * if no errors, else it returns appropriate errors.
+     * 
+     * @return bool|array
+     */
+    public function passes()
+    {
+        if (count($this->errors) > 0) {
```

Comment:
This bit of logic makes your validator stateful and hard to re-use over multiple executions: if the validator pushes all the results to an immutable value object, then we get a better/cleaner API